File: crates/flighthq-scene-gl/src/depth_gl_mesh_material_renderer.rs
//! The built-in depth forward renderer.
//!
//! Ports `@flighthq/scene-gl` `depthGlMeshMaterialRenderer.ts`.

use flighthq_materials::unlit_materials::depth_material_kind;
use flighthq_render_gl::GlRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};

use crate::gl_mesh_material_registry::{
    GlMeshMaterialRenderer, MeshMaterial, register_gl_mesh_material_renderer,
};
use crate::gl_scene_runtime::{GlMeshUpload, GlSceneRuntime};

/// The built-in depth renderer.
pub struct DepthGlMeshMaterialRenderer;

impl GlMeshMaterialRenderer for DepthGlMeshMaterialRenderer {
    fn bind(
        &self,
        _state: &mut GlRenderState,
        _scene: &mut GlSceneRuntime,
        _material: Option<&dyn MeshMaterial>,
        _lights: &SceneLightBlock,
        _camera: &Camera,
    ) {
        // Stub: requires the family prelude program + material uniform upload.
    }

    fn draw(
        &self,
        _state: &mut GlRenderState,
        _scene: &mut GlSceneRuntime,
        _proxy: &SceneRenderProxy,
        _upload: &GlMeshUpload,
    ) {
        // Stub: requires draw_gl_mesh_subset.
    }
}

/// Registers the built-in depth renderer.
pub fn register_depth_gl_material(scene: &mut GlSceneRuntime) {
    register_gl_mesh_material_renderer(
        scene,
        depth_material_kind(),
        Box::new(DepthGlMeshMaterialRenderer),
    );
}

#[cfg(test)]
mod tests {
    use super::*;
    use flighthq_materials::unlit_materials::depth_material_kind;

    use crate::gl_mesh_material_registry::get_gl_mesh_material_renderer;
    use crate::gl_scene_runtime::create_gl_scene_runtime;

    // register_depth_gl_material

    #[test]
    fn register_depth_gl_material_registers_for_the_depth_kind() {
        let mut scene = create_gl_scene_runtime();
        register_depth_gl_material(&mut scene);
        assert!(get_gl_mesh_material_renderer(&scene, depth_material_kind()).is_some());
    }
}

File: packages/scene-gl/src/depthGlMeshMaterialRenderer.ts
import type {
  Camera3D,
  DepthMaterial,
  GlMeshMaterialRenderer,
  GlRenderState,
  Material,
  MeshGeometry,
  Scene3DLightBlock,
  Scene3DRenderProxy,
} from '@flighthq/types';
import { DepthMaterialKind } from '@flighthq/types';

import { bindGlDebugRange, ensureGlDebugProgram } from './glDebugPrelude';
import { registerGlMeshMaterialRenderer } from './glMeshMaterialRegistry';
import { beginGlMeshDraw, drawGlMeshSubset, setGlMeshViewProjection } from './glMeshProgram';
import { getGlScene3DRuntime } from './glScene3DRuntime';

// The built-in Depth forward renderer (GlMeshMaterialRenderer for DepthMaterialKind). A lighting-
// independent debug/utility pass material: bind selects the debug program in depth mode, uploads the
// camera view-projection, and sets the material's [near, far] linearization range; draw issues the
// indexed draw. The fragment scene2d linearizes window-space depth into eye space and writes it as
// grayscale LINEAR color. Lights are ignored. See registerDepthGlMaterial to install it.
export const depthGlMeshMaterialRenderer: GlMeshMaterialRenderer = {
  bind(
    state: GlRenderState,
    material: Readonly<Material> | null,
    _lights: Readonly<Scene3DLightBlock>,
    camera: Readonly<Camera3D>,
  ): void {
    const gl = state.gl;
    const depth = material as Readonly<DepthMaterial> | null;
    const program = ensureGlDebugProgram(state, { hasNormalMap: false, mode: 'depth' });
    beginGlMeshDraw(state, program, depth !== null && depth.doubleSided);
    setGlMeshViewProjection(gl, program.locViewProjection, camera);

    if (depth === null) {
      bindGlDebugRange(state, program, 0, 1);
      return;
    }
    bindGlDebugRange(state, program, depth.near, depth.far);
  },

  draw(state: GlRenderState, proxy: Readonly<Scene3DRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    const program = getGlScene3DRuntime(state).activeMeshProgram;
    if (program === null) return;
    drawGlMeshSubset(state, program, proxy, geometry);
  },
};

// Registers the built-in Depth renderer for DepthMaterialKind on this state. Opt-in (no top-level
// side effect); call once per GlRenderState before drawScene3D so meshes with DepthMaterials draw.
export function registerDepthGlMaterial(state: GlRenderState): void {
  registerGlMeshMaterialRenderer(state, DepthMaterialKind, depthGlMeshMaterialRenderer);
}
